### DISCUSSION TITLE
Node cannot be copied

### DISCUSSION THREAD

**Discussion by Samsy** [donmccurdy/glTF-Transform#577] [General]
Hey hello ! 

I'm trying to understand the reason why a : " Node cannot be copied ".
Trying to read sources and understand but completely lost in there, but basically, what would prevent a clone() on a node  ?

**Comment by donmccurdy** [comment]
By default `.clone()` and `.copy()` are shallow, they don't make copies of resources that the target object refers to. That behavior doesn't work for Nodes and Scenes, because a Node can only have one parent. If the library shallow-cloned a Node with children, the scene would be left in an invalid state because its children would have two parents.

The easiest way around this would be to make a new node, and then copy the contents of the old one with a custom resolving function. For example, to deeply-clone the Node:

```js
const resolve = (source) => {
  if (source instanceof Node) {
    return document.createNode().copy(source, resolve);
  }
  return source;
};

const clonedNode = document.createNode().copy(originalNode, resolve);
```
